gx_suite_definitions/proteomics.ipynb

```diff
@@ -0,0 +1,276 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import synapseclient\n",
+    "\n",
+    "import great_expectations as gx\n",
+    "\n",
+    "context = gx.get_context(project_root_dir='../src/agoradatatools/great_expectations')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Create Expectation Suite for Proteomics Data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Get Example Data File"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "syn = synapseclient.Synapse()\n",
+    "syn.login()\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "proteomics_data_file = syn.get(\"syn18691125\").path\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create Validator Object on Data File"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "validator = context.sources.pandas_default.read_json(\n",
+    "    proteomics_data_file\n",
+    ")\n",
+    "validator.expectation_suite_name = \"proteomics\"\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Add Expectations to Validator Object For Each Column"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# uniqid\n",
+    "validator.expect_column_values_to_be_of_type(\"uniqid\", \"str\")\n",
+    "validator.expect_column_values_to_not_be_null(\"uniqid\")\n",
+    "validator.expect_column_values_to_match_regex(\"uniqid\", \"^^\\\\w{2,}|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}$\")"
```

Comment:
Can you specify what this regex matches? If I recall, protein IDs don't really follow a strict format like Ensembl IDs do.

Reply:
This regex string was pulled directly from the R rules [here](https://github.com/JessterB/agora-data-validation/blob/275dd46f48a69ec23da8a021a793da1036beb0b7/src/rules/rules_proteomics.R#L40).

I wasn't totally clear on what all this regex string meant, so I asked chatgpt and here is the breakdown it gave:

> ^ - Anchors the regex at the start of the string.
> \w{2,} - Matches any word character (alphanumeric + underscore) at least 2 times.
> | - Acts as an OR operator.
> [A-NR-Z][0-9] - Matches a single character that is between A-N or R-Z, followed by a digit (0-9).
> ([A-Z][A-Z0-9]{2}[0-9]){1,2} - Matches one or two occurrences of:
> [A-Z] - A single uppercase letter.
> [A-Z0-9]{2} - Two characters that are either uppercase letters or digits.
> [0-9] - A single digit.
> $ - Anchors the regex at the end of the string.
> Putting it all together, this regex matches a string that either:
> 
> Starts with at least two word characters (letters, digits, or underscores).
> Starts with a letter from A to N or R to Z, followed by a digit, and then one or two sequences of an uppercase letter followed by two alphanumeric characters and ending with a digit.

Reply:
Ok, maybe we can let @JessterB weigh in. This feels really fragile. From what I can tell looking at the file for this dataset, the uniqueid should generally be `<gene_symbol>|<uniprotid>`, so if anything the regex should be `<regex_for_hgnc_symbol>\|<regex_for_uniprotid>`, right...?

Reply:
Now that I am looking at it closer, I think the AI may have been wrong in interpreting the `|` as an OR operator and instead it is acting as the pipe in the `^<regex_for_hgnc_symbol>\|<regex_for_uniprotid>$` format that you suggested.

Reply:
I'm still waiting to see what @JessterB thinks about this. Two things I just noticed: 
1. The pipe character needs to be escaped in this regex because it's literal, not an 'or' operator
2. I also just noticed another issue with this. Since this should be `<gene_symbol>|<uniprotid>`, the regex for the `uniprotid` field should match the back half of this regex, and it doesn't. For `uniprotid` it's just `^[a-zA-Z0-9.-]*$`.  In Jess's validation file [here](https://github.com/JessterB/agora-data-validation/blob/master/src/rules/rules_proteomics.R), the regex for `uniprotid` is listed as `[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}`, which comes directly from UniProt. So I think that needs to be the regex for the `uniprotid` field here. It looks like only half of this regex is currently included in the `uniqid` but I don't understand why the whole thing isn't.

Reply:
I can't seem to be able to reply to one of Jess's comments below, so I'm putting it in this thread.

From @JessterB:
> Since uniqID is just hgnc|uniprot and we have regexes for both of those fields, I see a couple options here:
> 1. Assume we concatenate these strings, which are independently validated via the two regexes, correctly
> 2. Validate this value with a simple regex that checks for two generic strings with a pipe between them
> 3. Create a regex that leverages both of the other regexes to validate the first string, the pipe, and the second string

I would lean toward option 2 or 3. I'm keeping in mind the SRM proteomics data, which I had to programatically look up UniProt IDs for and create a `uniqid` based off that. So it'd be good to have a double-check that this field is as expected across all proteomics files. I think option 2 would be for simplicity/easier to maintain (maybe?) and option 3 would be for thoroughness, but either way it'd be good to put a comment in the code here explaining what this regex is for. @BWMac what do you think?

Reply:
I think we should try for option 3 because it is the most robust. I will see how that goes and then if it works I will apply it to this PR and the `srm` one!

Reply:
There is one `uniqid` value `CON__P35908` which does not match the pattern at all. The full record looks like:
```
  {
    "uniqid": "CON__P35908",
    "hgnc_symbol": "CON__P35908",
    "uniprotid": "P35908",
    "ensembl_gene_id": "ENSG00000172867",
    "tissue": "AntPFC",
    "log2_fc": null,
    "ci_upr": null,
    "ci_lwr": null,
    "pval": null,
    "cor_pval": null
  }
```
And it appears as though the concatenation is not happening correctly.

Reply:
Weird, according to [ensembl](https://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000172867;r=12:52644558-52652211) and [Agora](https://agora-develop.adknowledgeportal.org/genes/ENSG00000172867), that ENSG maps to KRT2. Where does CON__P35908 come from?

Reply:
Not sure... We do have records for `KRT2` like 
```
  {
    "uniqid": "KRT2|P35908",
    "hgnc_symbol": "KRT2",
    "uniprotid": "P35908",
    "ensembl_gene_id": "ENSG00000172867",
    "tissue": "MFG",
    "log2_fc": null,
    "ci_upr": null,
    "ci_lwr": null,
    "pval": null,
    "cor_pval": null
  }
```

Reply:
OK so this regex string is what I came up with:
```
^(?:[A-Za-z0-9_.-]+\\|[A-Za-z0-9_.-]+|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$
```
Which enforces the `<gene_symbol>|<uniprotid>` pattern making sure that the `|` character is there, so the `CON__P35908` ones are the only ones failing right now.
```
{
  "success": false,
  "result": {
    "element_count": 22704,
    "unexpected_count": 4,
    "unexpected_percent": 0.017618040873854827,
    "partial_unexpected_list": [
      "CON__P35908",
      "CON__P35908",
      "CON__P35908",
      "CON__P35908"
    ],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.017618040873854827,
    "unexpected_percent_nonmissing": 0.017618040873854827
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}
```

Reply:
@JessterB I see that `CON__P35908` record in the input [dataset](https://www.synapse.org/Synapse:syn18689335.3), but that's sort of where the scope of my understanding ends. Is this something that we want to allow for now? Or does there need to be a change to the input data before we implement this GX suite?

Reply:
This issue originates in the source file, where we have results for the same ENSG under the two different hgnc values. The concatenated uniqid value is wrong in the source file:
![Screen Shot 2024-07-23 at 11 47 53 AM](https://github.com/user-attachments/assets/42f75ede-99bc-4647-9377-c517a558b59d)

Options:
1. Create a new version of the source file where we correct the weird hgnc and uniqid values on the CON__ records to align with the KRT2 records
2. Add a custom transform for proteomics that concatenates the UniProtID with the hgnc that we resolve for the ENSG from biomart via gene_info
3. Add a custom transform for proteomics that just cleans up this specific record as outlined in 1

@jaclynbeck-sage any opinions or other ideas?

Reply:
@BWMac This is something we should probably fix either in the source, or during data processing. I'm thinking we should hold off on deciding what to do with this GX suite until Jaclyn and I have decided how to address this. Depending on what we decide, and how long we think it's going to take us, it should become clear whether it's worth changing the GX suite to tolerate the issue or not. Does that work for you?

Reply:
Sure, no rush on my end! 

My two cents is that fixing this in the source data is probably preferable (unless there is some reason I am unaware of that makes this untenable), but happy to move forward with whatever you and Jaclyn determine is best!.

Reply:
I agree we either need to fix it at the source or during transforms. This is odd because "CON_P35908" isn't a valid HGNC symbol either. I actually don't think underscores exist in HGNC symbols? Lemme go check on that...

Reply:
Oh I guess underscores do exist in HGNC symbols. Regardless, that particular one isn't an HGNC symbol.

Reply:
@JessterB do you know where the source file originates from? Did someone at Sage make it or was it handed to us by an external contributor?

Reply:
@jaclynbeck-sage the v3 source file we're using is from 2019, which is before my time. It looks like JakeG and Anna added the first 3 versions of it to synapse; the only [provenance](https://www.synapse.org/Synapse:syn18689335.1) is on v1, and while it isn't particularly helpful it suggests that Jake probably generated it.

There's also now a v4, but as far as I can tell it's identical to v3 so I'm not sure why it even exists.

Reply:
@jaclynbeck-sage Actually, I'm pretty sure that Jake harmonized the results from multiple studies for the LFQ proteomics, as DLPFC is from [Banner](https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage/StudyData?Study=syn7170616) and AntPFC is from [MSBB](https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage/StudyDetails?Study=syn3159438).

Reply:
Ok, let me do a little sleuthing and see what those original files look like.

Reply:
@BWMac The invalid rows will be filtered out via https://sagebionetworks.jira.com/browse/AG-1526; let's sit on this PR until we've made that update. I set that ticket as a blocker on IBCDPE-794.